### ISSUE TITLE
ple 实现有问题

### ISSUE THREAD

**Issue by wolaiye1010** [easezyc/Multitask-Recommendation-Library#4]
task_fea = [emb for i in range(self.task_num + 1)]
        for i in range(self.layers_num):
            for j in range(self.task_num + 1):
                fea = torch.cat([self.expert[i][index](task_fea[j]).unsqueeze(1) for index in self.task_expert_index[j]], dim = 1)
#上面这一行代码会导致，shareExpert 的输入是上一层的task的输出，第一层的时候没有问题，第二层就会有问题，如下图红线所示。

                gate_value = self.gate[i][j](task_fea[j]).unsqueeze(1)
                task_fea[j] = torch.bmm(gate_value, fea).squeeze(1)

![image](https://user-images.githubusercontent.com/15831945/191630739-42de606f-099f-4424-81dc-134022982be9.png)

**Comment by easezyc** [comment]
不太明白你说的是什么问题，shareExpert的输入不就应该是上一层所有shared expert以及specific expert的输出吗

**Comment by wolaiye1010** [comment]
@easezyc , 我描述有问题，实际的问题点就是我红线画的，第一层的每个specificExpert的输出，应该只输入到相应的下一层的specific task，share expert的输出到share expert，但现在的实现变成了我红线所画的流程。

**Comment by wolaiye1010** [comment]
@easezyc ，我可以改下，发个pr你看下

**Comment by easezyc** [comment]
感觉是你理解错了，红线才是对的，你看这个代码，也是这种实现，不过他只有一层
https://github.com/huangjunheng/recommendation_model/blob/master/mmoe_model/PLE.py

**Comment by easezyc** [comment]
It is worth noting that the selected matrix of **the shared module in PLE is slightly different from task-specific modules, as it consists of all shared experts and task-specific experts in this layer**.
论文原文

**Comment by wolaiye1010** [comment]
@easezyc ,"It is worth noting that the selected matrix of the shared module in PLE is slightly different from task-specific modules, as it consists of all shared experts and task-specific experts in this layer."
这句话的意思不是红线相连，而是shareExpert 的输出是由shareExpert 的gate混合 上一层 所有taskSpecific +expert的输出

**Comment by wolaiye1010** [comment]
@easezyc ， 可以看下这个 高star的tf的开源实现 https://github.com/shenweichen/DeepCTR/blob/master/deepctr/models/multitask/ple.py
是没有红线部分的，每个task的输入就是上一层的混合后 的task的输出，而且论文里的结构图片也是这么画的

**Comment by wolaiye1010** [comment]
> 感觉是你理解错了，红线才是对的，你看这个代码，也是这种实现，不过他只有一层 https://github.com/huangjunheng/recommendation_model/blob/master/mmoe_model/PLE.py

这个代码只实现了单层的cgc 网络，没有实现后面的层。

**Comment by easezyc** [comment]
@wolaiye1010 
我的代码实现的不就是这个原理吗，我看不懂你在说我代码到底哪的问题，你的代码写出来实现的跟我一个逻辑，并且行数更多
下一层shared expert的输入是上一层所有expert的输出乘以gate值聚合（那句英文的含义，没看懂我哪里理解错了）
下一层specific的输入是上一层对应specific的输出以及shared输出乘以gate值聚合
这个tf的代码我看了，跟我实现的原理一样啊
我的代码唯一的缺陷是多算了一步，最后一层输出不用算shared这块，解决也非常简单，简单修改如下
for i in range(self.layers_num):
            count = self.task_num + 1 if i < self.layers_num - 1 else self.task_num
            for j in range(count):
                fea = torch.cat([self.expert[i][index](task_fea[j]).unsqueeze(1) for index in self.task_expert_index[j]], dim = 1)
                gate_value = self.gate[i][j](task_fea[j]).unsqueeze(1)
                task_fea[j] = torch.bmm(gate_value, fea).squeeze(1)

**Comment by easezyc** [comment]
@wolaiye1010 
我get了我的代码的问题在哪了，你的commit我已经merge进去了，感谢你的贡献

**Comment by easezyc** [comment]
@wolaiye1010 
也麻烦你再跑一下你修改的这个PLE是否能跑通

**Comment by wolaiye1010** [comment]
@easezyc , 已测试没有问题，pull request里有test方法

**Comment by easezyc** [comment]
@wolaiye1010 
也再麻烦你确认一下你提交的代码是否能跑通我给的这几个数据集（我最近没什么时间来测试），看看这个表格里的实验结果（3次重复实验）是否需要更正一下，再次感谢你的贡献
![1664169309597](https://user-images.githubusercontent.com/20279673/192198720-c4d849c2-e5af-44da-8bdc-46e4667c276c.png)
